File: src/sys/java/fan/sys/Zip.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   18 Aug 06  Brian Frank  Creation
//
package fan.sys;

import java.util.Enumeration;
import java.util.zip.ZipFile;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;
import java.util.zip.ZipOutputStream;
import java.util.zip.GZIPOutputStream;
import java.util.zip.GZIPInputStream;
import java.util.zip.Deflater;
import java.util.zip.DeflaterOutputStream;
import java.util.zip.Inflater;
import java.util.zip.InflaterInputStream;

/**
 * Zip is used to read/write compressed zip files and streams.
 */
public final class Zip
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static Zip open(File file) { return new Zip(file); }

  private Zip(File f)
  {
    try
    {
      // only support local files
      if (!(f instanceof LocalFile))
        throw IOErr.make("Only local files supported: " + f);

      // open the file
      this.file = (LocalFile)f;
      this.zipFile = new ZipFile(file.file);
    }
    catch (java.io.IOException e)
    {
      throw IOErr.make(e);
    }
  }

  public static Zip read(InStream in) { return new Zip(in); }

  private Zip(InStream in)
  {
    this.zipIn = new ZipInputStream(SysInStream.java(in));
  }

  public static Zip write(OutStream out) { return new Zip(out); }

  private Zip(OutStream out)
  {
    this.zipOut = new ZipOutputStream(SysOutStream.java(out));
  }

//////////////////////////////////////////////////////////////////////////
// Obj
//////////////////////////////////////////////////////////////////////////

  public Type typeof() { return Sys.ZipType; }

  public String toStr()
  {
    if (file != null) return file.toStr();
    return super.toStr();
  }

//////////////////////////////////////////////////////////////////////////
// Access
//////////////////////////////////////////////////////////////////////////

  public File file()
  {
    return file;
  }

  public Map<Uri,File> contents()
  {
    if (contents == null)
    {
      if (zipFile == null) return null;
      contents = contents(zipFile);
    }
    return contents.ro();
  }

  public static Map<Uri,File> contents(ZipFile zipFile)
  {
    Map c = new Map(Sys.UriType, Sys.FileType);
    Enumeration e = zipFile.entries();
    while (e.hasMoreElements())
    {
      ZipEntry entry = (ZipEntry)e.nextElement();
      ZipEntryFile f = new ZipEntryFile(zipFile, entry);
      c.set(f.uri, f);
    }
    return c;
  }

  public File readNext()
  {
    if (zipIn == null) throw UnsupportedErr.make("Zip not opened for reading");
    try
    {
      ZipEntry entry = zipIn.getNextEntry();
      if (entry == null) return null;
      return new ZipEntryFile(this, entry);
    }
    catch (java.io.IOException e)
    {
      throw IOErr.make(e);
    }
  }

  public void readEach(Func f)
  {
    File file;
    while ((file = readNext()) != null)
    {
      f.call(file);
    }
  }

  public OutStream writeNext(Uri path) { return writeNext(path, DateTime.now(), null); }
  public OutStream writeNext(Uri path, DateTime modifyTime) { return writeNext(path, modifyTime, null); }
  public OutStream writeNext(Uri path, DateTime modifyTime, Map opts)
  {
    if (zipOut == null) throw UnsupportedErr.make("Zip not opened for writing");
    if (path.frag() != null) throw ArgErr.make("Path must not contain fragment: " + path);
    if (path.queryStr() != null) throw ArgErr.make("Path must not contain query: " + path);

    // Java 1.7+ supports ZIP64 which supports over 65,535 files, but
    // previous versions silently fail which is really bad; so add
    // Fantom specific sanity check here
    if (Sys.javaVersion < Sys.JAVA_1_7)
    {
      if (zipOutCount >= 65535) throw UnsupportedErr.make("Zip cannot handle more than 65535 files");
      zipOutCount++;
    }

    try
    {
      // init ZipEntry
      String zipPath = path.toString();
      if (zipPath.startsWith("/")) zipPath = zipPath.substring(1);
      ZipEntry entry = new ZipEntry(zipPath);

      // options
      entry.setTime(modifyTime.toJava());
      if (opts != null)
      {
        if (opts.get("level") != null)
        {
          int level = ((Long)opts.get("level")).intValue();
          zipOut.setLevel(level);
          zipOut.setMethod(level == 0 ? ZipOutputStream.STORED : ZipOutputStream.DEFLATED);
        }
        if (opts.get("comment")          != null) entry.setComment((String)opts.get("comment"));
        if (opts.get("crc")              != null) entry.setCrc((Long)opts.get("crc"));
        if (opts.get("extra")            != null) entry.setExtra(((Buf)opts.get("extra")).safeArray());
        if (opts.get("compressedSize")   != null) entry.setCompressedSize((Long)opts.get("compressedSize"));
        if (opts.get("uncompressedSize") != null) entry.setSize((Long)opts.get("uncompressedSize"));
      }

      // putNextEntry and return as SysOutStream
      zipOut.putNextEntry(entry);
      return new SysOutStream(zipOut)
      {
        public boolean close()
        {
          try
          {
            zipOut.closeEntry();
            return true;
          }
          catch (java.io.IOException e)
          {
            e.printStackTrace();
            return false;
          }
        }
      };
    }
    catch (java.io.IOException e)
    {
      throw IOErr.make(e);
    }
  }

  public boolean finish()
  {
    if (zipOut == null) throw UnsupportedErr.make("Zip not opened for writing");
    try
    {
      zipOut.finish();
      return true;
    }
    catch (java.io.IOException e)
    {
      e.printStackTrace();
      return false;
    }
  }

  public boolean close()
  {
    try
    {
      if (zipFile != null) zipFile.close();
      if (zipIn != null)   zipIn.close();
      if (zipOut != null)  zipOut.close();
      return true;
    }
    catch (java.io.IOException e)
    {
      e.printStackTrace();
      return false;
    }
  }

  public static long unzipInto(File zipFile, File dir)
  {
    if (!dir.isDir()) throw ArgErr.make("Not dir: " + dir);
    Zip zip = null;
    try
    {
      int count = 0;
      zip = read(zipFile.in());
      File entry;
      while ((entry = zip.readNext()) != null)
      {
        String relUri = entry.uri().toStr().substring(1);
        File dest = dir.plus(Uri.fromStr(relUri));
        if (entry.isDir()) { dest.create(); continue; }
        OutStream out = dest.out();
        try
        {
          entry.in().pipe(out);
        }
        finally
        {
          out.close();
        }
        if (entry.modified() != null) dest.modified(entry.modified());
        count++;
      }
      return count;
    }
    finally
    {
      if (zip != null) zip.close();
    }
  }

//////////////////////////////////////////////////////////////////////////
// GZIP
//////////////////////////////////////////////////////////////////////////

  public static OutStream gzipOutStream(OutStream out)
  {
    try
    {
      return new SysOutStream(new GZIPOutputStream(SysOutStream.java(out)));
    }
    catch (java.io.IOException e)
    {
      throw IOErr.make(e);
    }
  }

  public static InStream gzipInStream(InStream in)
  {
    try
    {
      return new SysInStream(new GZIPInputStream(SysInStream.java(in)));
    }
    catch (java.io.IOException e)
    {
      throw IOErr.make(e);
    }
  }

//////////////////////////////////////////////////////////////////////////
// Deflate/Inflate
//////////////////////////////////////////////////////////////////////////

  public static OutStream deflateOutStream(OutStream out) { return deflateOutStream(out, null); }
  public static OutStream deflateOutStream(OutStream out, Map opts)
  {
    int level = Deflater.DEFAULT_COMPRESSION;
    boolean nowrap = false;
    if (opts != null)
    {
      if (opts.get("nowrap") != null) nowrap = ((Boolean)opts.get("nowrap")).booleanValue();
      if (opts.get("level") != null) level = ((Long)opts.get("level")).intValue();
    }
    Deflater d = new Deflater(level, nowrap);
    return new SysOutStream(new DeflaterOutputStream(SysOutStream.java(out), d));
  }

  public static InStream deflateInStream(InStream in) { return deflateInStream(in, null); }
  public static InStream deflateInStream(InStream in, Map opts)
  {
    boolean nowrap = false;
    if (opts != null)
    {
      if (opts.get("nowrap") != null) nowrap = ((Boolean)opts.get("nowrap")).booleanValue();
    }
    Inflater i = new Inflater(nowrap);
    return new SysInStream(new InflaterInputStream(SysInStream.java(in), i));
  }


//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  LocalFile file;           // open only
  ZipFile zipFile;          // open only
  Map contents;             // open only
  ZipInputStream zipIn;     // read only
  ZipOutputStream zipOut;   // write only
  int zipOutCount;          // write only

}

File: src/sys/dotnet/fan/sys/Zip.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   18 Apr 07  Andy Frank  Creation
//

using System.Collections;
using ICSharpCode.SharpZipLib.Zip;

namespace Fan.Sys
{
  /// <summary>
  /// Zip.
  /// </summary>
  public sealed class Zip : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Construction
  //////////////////////////////////////////////////////////////////////////

    public static Zip open(File file) { return new Zip(file); }

    private Zip(File f)
    {
      try
      {
        // only support local files
        if (!(f is LocalFile))
          throw IOErr.make("Only local files supported: " + f).val;

        // open the file
        this.m_file = (LocalFile)f;
        this.m_zipFile = new ZipFile(m_file.m_file.FullName);
      }
      catch (System.Exception e)
      {
        // NOTE: use ctor instead of make() to force type == IOErr
        throw new IOErr(e).val;
      }
    }

    public static Zip read(InStream ins) { return new Zip(ins); }

    private Zip(InStream ins)
    {
      this.m_zipIn = new ZipInputStream(SysInStream.dotnet(ins));
    }

    public static Zip write(OutStream outs) { return new Zip(outs); }

    private Zip(OutStream outs)
    {
      this.m_zipOut = new ZipOutputStream(SysOutStream.dotnet(outs));
    }

  //////////////////////////////////////////////////////////////////////////
  // Obj
  //////////////////////////////////////////////////////////////////////////

    public override Type @typeof() { return Sys.ZipType; }

    public override string toStr()
    {
      if (m_file != null) return m_file.toStr();
      return base.toStr();
    }

  //////////////////////////////////////////////////////////////////////////
  // Access
  //////////////////////////////////////////////////////////////////////////

    public File file()
    {
      return m_file;
    }

    public Map contents()
    {
      if (m_contents == null)
      {
        if (m_zipFile == null) return null;
        m_contents = contents(m_zipFile);
      }
      return m_contents.ro();
    }

    public static Map contents(ZipFile zipFile)
    {
      Map c = new Map(Sys.UriType, Sys.FileType);
      IEnumerator e = zipFile.GetEnumerator();
      while (e.MoveNext())
      {
        ZipEntry entry = (ZipEntry)e.Current;
        ZipEntryFile f = new ZipEntryFile(zipFile, entry);
        c.set(f.m_uri, f);
      }
      return c;
    }

    public File readNext()
    {
      if (m_zipIn == null) throw UnsupportedErr.make("Zip not opened for reading").val;
      try
      {
        ZipEntry entry = m_zipIn.GetNextEntry();
        if (entry == null) return null;
        return new ZipEntryFile(this, entry);
      }
      catch (System.IO.IOException e)
      {
        throw IOErr.make(e).val;
      }
    }

    public OutStream writeNext(Uri path) { return writeNext(path, DateTime.now()); }
    public OutStream writeNext(Uri path, DateTime modifyTime)
    {
      if (m_zipOut == null) throw UnsupportedErr.make("Zip not opened for writing").val;
      if (path.frag() != null) throw ArgErr.make("Path must not contain fragment: " + path).val;
      if (path.queryStr() != null) throw ArgErr.make("Path must not contain query: " + path).val;
      try
      {
        string zipPath = path.ToString();
        if (zipPath.StartsWith("/")) zipPath = zipPath.Substring(1);
        ZipEntry entry = new ZipEntry(zipPath);
        entry.DateTime = new System.DateTime(modifyTime.dotnet());
        m_zipOut.PutNextEntry(entry);
        return new ZipSysOutStream(m_zipOut);
      }
      catch (System.IO.IOException e)
      {
        throw IOErr.make(e).val;
      }
    }

    public bool finish()
    {
      if (m_zipOut == null) throw UnsupportedErr.make("Zip not opened for writing").val;
      try
      {
        m_zipOut.Finish();
        return true;
      }
      catch (System.IO.IOException)
      {
        return false;
      }
    }

    public bool close()
    {
      try
      {
        if (m_zipFile != null) m_zipFile.Close();
        if (m_zipIn != null)   m_zipIn.Close();
        if (m_zipOut != null)  m_zipOut.Close();
        return true;
      }
      catch (System.IO.IOException)
      {
        return false;
      }
    }

  //////////////////////////////////////////////////////////////////////////
  // ZipSysOutStream
  //////////////////////////////////////////////////////////////////////////

    internal class ZipSysOutStream : SysOutStream
    {
      public ZipSysOutStream(ZipOutputStream zout) : base(zout)
      {
        this.zout = zout;
      }
      public override bool close()
      {
        try
        {
          zout.CloseEntry();
          return true;
        }
        catch (System.IO.IOException)
        {
          return false;
        }
      }
      private ZipOutputStream zout;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    internal LocalFile m_file;           // open only
    internal ZipFile m_zipFile;          // open only
    internal Map m_contents;             // open only
    internal ZipInputStream m_zipIn;     // read only
    internal ZipOutputStream m_zipOut;   // write only

  }
}